### ISSUE TITLE
/(?{ code })/ construct suffers from "doesn't stay shared" problem

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4676]
Migrated from [rt.perl.org#8030](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=8030) (status was 'resolved')

Searchable as RT8030$

**Comment by p5pRT** [comment]
### From john@pc.xs4all.nl

#### Created by john@pc.xs4all.nl


This is a bug report for perl from john@&#8203;pc\.xs4all\.nl\,
generated with the help of perlbug 1\.33 running under perl v5\.7\.2\.


\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
The code executed inside a regular expression\, using the \(?\{ code \}\)
construct\, has the same "doesn't stay shared" problem as nesting a
named subroutine in a subroutine\. Except that no warning is generated\.

Here is an example program that displays the problem&#8203;:

\-\-\-\-8\<\-\-\-BEGIN\-EXAMPLE\-\-\-\-\-
\#\!/usr/bin/perl \-w

use strict;

my $foo;

sub reg \{
&nbsp;    my $str = shift;

&nbsp;    my $bar;

&nbsp;    $str =~ m\{
&nbsp;    ^\(\[a\-z\]\+\)
&nbsp;    \(\\d\+\)
&nbsp;    \(?\{ $foo = $1; $bar = $2; \}\)
&nbsp;    \}x;

&nbsp;    return $bar;
\}

my $bar1 = reg\("abcd123"\);
print "bar1&#8203;: '$bar1'\, foo1&#8203;: '$foo'\\n";
my $bar2 = reg\("defg789"\);
print "bar2&#8203;: '$bar2'\, foo2&#8203;: '$foo'\\n";
\-\-\-\-END\-EXAMPLE\-\-\->8\-\-\-\-

This is the output of this script&#8203;:

bar1&#8203;: '123'\, foo1&#8203;: 'abcd'
Use of uninitialized value in concatenation \(\.\) or string at doesntstayshared line 24\.
bar2&#8203;: ''\, foo2&#8203;: 'defg'

The output is identical for perl versions perl5\.6\.1\, perl5\.7\.2 and yesterday's
bleadperl\.

I would have proposed a patch to issue a warning when this happens\, but I'm
not quite fluent enough in the internals to make it happen on short notice\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.7.2:

Configured by john at Sun Dec  9 23:05:55 CET 2001.

Summary of my perl5 (revision 5.0 version 7 subversion 2 patch 13539) configuration:
  Platform:
    osname=linux, osvers=2.2.19, archname=i686-linux
    uname='linux pc.xs4all.nl 2.2.19 #1 sun jun 24 15:06:07 cest 2001 i686 unknown '
    config_args='-Doptimize=-g -Dusedevel -des'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=define
  Compiler:
    cc='cc', ccflags ='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-g',
    cppflags='-DDEBUGGING -fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='egcs-2.91.66 19990314/Linux (egcs-1.1.2 release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lposix -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.3.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    DEVEL13539


@INC for perl v5.7.2:
    lib
    /usr/local/lib/perl5/5.7.2/i686-linux
    /usr/local/lib/perl5/5.7.2
    /usr/local/lib/perl5/site_perl/5.7.2/i686-linux
    /usr/local/lib/perl5/site_perl/5.7.2
    /usr/local/lib/perl5/site_perl
    .


Environment for perl v5.7.2:
    HOME=/usr/home/john
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/home/john/bin:/bin:/usr/bin:/sbin:/usr/sbin:/usr/local/bin:/usr/X11R6/bin:/usr/games:/usr/local/sbin:/usr/local/pgsql/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash


```
</details>

**Comment by p5pRT** [comment]
### From @vanstyn

Jan\-Pieter Cornet \<john@&#8203;pc\.xs4all\.nl> wrote&#8203;:
:The code executed inside a regular expression\, using the \(?\{ code \}\)
:construct\, has the same "doesn't stay shared" problem as nesting a
:named subroutine in a subroutine\. Except that no warning is generated\.
\[snip\]
:I would have proposed a patch to issue a warning when this happens\, but I'm
:not quite fluent enough in the internals to make it happen on short notice\.

I think we don't yet know what it \_should\_ do\. Once we know that\,
hopefully we can fix it either to DWYM or give the appropriate warning\.

Hugo

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

On Mon\, Dec 10\, 2001 at 03&#8203;:47&#8203;:24PM \+0000\, Hugo van der Sanden wrote&#8203;:
> Jan\-Pieter Cornet \<john@&#8203;pc\.xs4all\.nl> wrote&#8203;:
> :The code executed inside a regular expression\, using the \(?\{ code \}\)
> :construct\, has the same "doesn't stay shared" problem as nesting a
> :named subroutine in a subroutine\. Except that no warning is generated\.
> \[snip\]
> :I would have proposed a patch to issue a warning when this happens\, but I'm
> :not quite fluent enough in the internals to make it happen on short notice\.
> 
> I think we don't yet know what it \_should\_ do\. Once we know that\,
> hopefully we can fix it either to DWYM or give the appropriate warning\.

Oops\, sorry for not being clearer\. The problem is that the lexical variables
from the scope of the enclosing subroutine\, are only visible the first
time the code in the \(?\{ code\.\.\. \}\) block executes\. On the second and
all subsequent invocations\, the lexicals in the regular expression code
block no longer use the pad variables of the current function\.

This "one" liner makes it even clearer\, I hope&#8203;:

$ perl \-le 'sub e\{my $x; shift =~ /\(\.\*\)\(?\{$x = $1\}\)/; \\$x\} 
$re1 = e\("JAPH"\);  print "1&#8203;: $$re1"; 
$re2 = e\("JAPH2"\); print "2&#8203;: $$re2"; print "1 again&#8203;: $$re1"'
1&#8203;: JAPH
2&#8203;: 
1 again&#8203;: JAPH2

\(output is identical for perl 5\.6\.1\, 5\.7\.2 and bleadperl "DEVEL13539"\)

The regular expression matches a string\, and assigns it to the lexical $x\.

This works the first time\. The second time\, it fails to assign the current
lexical $x\. Instead\, it assigns it to where $x was the first time the
subroutine was called\. Which is still preserved because of the returned
reference\.

I recognised this behaviour as the one that the warning "Variable \\"%s\\"
will not stay shared" describes in perldiag\. However\, in the named subroutine
case\, perl gives a warning&#8203;:

$ perl \-we 'sub a\{my$x; sub b\{$x\+\+\}\}'
Variable "$x" will not stay shared at \-e line 1\.

I suspect that exactly the same thing is happening \(since the regex is
only compiled once\, the pad references are only resolved once\. I guess\)\.

Ah\, yes\. If we introduce an interpolated string \(that actually changes\)\,
things work fine&#8203;:

$ perl \-le 'sub e\{my$x; use re "eval"; shift =~ /\(?\!\[^\\s\\S\]$a\)\(\.\*\)\(?\{$x = $1\}\)/;
$a\+\+; \\$x\}
$re1 = e\("JAPH"\); print "1&#8203;: $$re1";
$re2 = e\("JAPH2"\); print "2&#8203;: $$re2"; print "1 again&#8203;: $$re1"'
1&#8203;: JAPH
2&#8203;: JAPH2
1 again&#8203;: JAPH

Is there some flag on a compiled regex that says "compile me again next time"?
That'd fix it \(at a performance penalty\, obviously\)\.

But like I suggested\, the "other" fix would be to issue the same warning
as the nested\-named subroutine does\.

Hope this helps\,

\-\- 
\#\!perl \-pl    \# This kenny\-filter is virus\-free as long as you don't copy it
$p=3\-2\*/\[^\\W\\dmpf\_\]/i;s\.\[a\-z\]\{$p\}\.vec\($f=join\(''\,$p\-1?chr\(sub\{$\_\[0\]\*9\+$\_\[1\]\*3\+
$\_\[2\]\}\->\(map\{/p|f/i\+/f/i\}split//\,$&\)\+97\)&#8203;:\('m'\,p\,f\)\[map\{\(\(ord$&\)%32\-1\)/$\_%3\}\(9\,
3\,1\)\]\)\,5\,1\)='\`'lt$&;$f\.eig;                   \# Jan\-Pieter Cornet

**Comment by p5pRT** [comment]
### From @cpansprout

This has been fixed by the commits leading up to eb58a7e122\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
